slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/DbSlashingProtection.java

```diff
@@ -43,16 +68,95 @@ public boolean maySignAttestation(
       final Bytes signingRoot,
       final UInt64 sourceEpoch,
       final UInt64 targetEpoch) {
-    return true;
+    final Optional<Long> validatorId = Optional.ofNullable(registeredValidators.get(publicKey));
+    if (validatorId.isEmpty()) {
+      LOG.error("Unregistered Eth2 validator for {}", publicKey);
+      return false;
+    }
+
+    if (sourceEpoch.compareTo(targetEpoch) > 0) {
+      LOG.error(
+          "Eth2 slashing protection for {} detected sourceEpoch {} greater than targetEpoch {}",
+          publicKey,
+          sourceEpoch,
+          targetEpoch);
+      return false;
+    }
+
+    return jdbi.inTransaction(
+        h -> {
+          final long id = validatorId.get();
+          final boolean maySign =
+              maySignAttestation(h, signingRoot, sourceEpoch, targetEpoch, id, publicKey);
+          if (maySign) {
+            final SignedAttestation signedAttestation =
+                new SignedAttestation(id, sourceEpoch, targetEpoch, signingRoot);
+            signedAttestationsDao.insertAttestation(h, signedAttestation);
+          }
+          return maySign;
+        });
+  }
+
+  private Boolean maySignAttestation(
+      final Handle handle,
+      final Bytes signingRoot,
+      final UInt64 sourceEpoch,
+      final UInt64 targetEpoch,
+      final long id,
+      final Bytes publicKey) {
+    // check for double vote, an existing attestation with same target epoch by different
+    // signing root
+    final Optional<SignedAttestation> existingAttestation =
+        signedAttestationsDao.findExistingAttestation(handle, id, targetEpoch);
+    if (existingAttestation.isPresent()) {
+      if (existingAttestation.get().getSigningRoot().equals(signingRoot)) {
+        return true;
+      } else {
+        LOG.error(
+            "Eth2 slashing protection for validator {} detected double signed attestation {}",
+            publicKey,
+            existingAttestation.get());
+        return false;
+      }
+    }
+
+    // check that no previous vote is surrounding the attestation
+    final Optional<SignedAttestation> surroundingAttestation =
+        signedAttestationsDao.findSurroundingAttestation(handle, id, sourceEpoch, targetEpoch);
+    if (surroundingAttestation.isPresent()) {
+      LOG.error(
+          "Eth2 slashing protection for validator {} detected surrounding attestation {} for attestation signingRoot={} sourceEpoch={} targetEpoch={}",
+          publicKey,
+          surroundingAttestation.get(),
+          signingRoot,
+          sourceEpoch,
+          targetEpoch);
+      return false;
+    }
+
+    // check that no previous vote is surrounded by attestation
+    final Optional<SignedAttestation> surroundedAttestation =
```

Comment:
could/should this be a separate function? A well named function means the comment could be removed

Reply:
don't think so, the logic is really only two lines. I think comment could be removed too the function call and the variable name make it obvious what the check is.